v1_20_R1/src/main/java/io/th0rgal/oraxen/nms/v1_20_R1/NMSHandler.java

```diff
@@ -203,4 +208,25 @@ public void applyMiningEffect(Player player) {
     public void removeMiningEffect(Player player) {
         ((CraftPlayer) player).getHandle().connection.send(new ClientboundRemoveMobEffectPacket(player.getEntityId(), MobEffects.DIG_SLOWDOWN));
     }
+
+    @Override
+    public String getNoteBlockInstrument(Block block) {
+        Material aboveMaterial = block.getRelative(BlockFace.UP).getType();
+        Block blockBelow = block.getRelative(BlockFace.DOWN);
+        BlockState blockStateBelow = ((CraftBlock) blockBelow).getNMS();
+
+        return switch (aboveMaterial) {
+            case SKELETON_SKULL -> "block.note_block.imitate.skeleton";
+            case PIGLIN_HEAD -> "block.note_block.imitate.piglin";
+            case ZOMBIE_HEAD -> "block.note_block.imitate.zombie";
+            case CREEPER_HEAD -> "block.note_block.imitate.creeper";
+            case DRAGON_HEAD -> "block.note_block.imitate.ender_dragon";
+            case WITHER_SKELETON_SKULL -> "block.note_block.imitate.wither_skeleton";
+            default -> {
+                CustomBlockMechanic customBlockMechanic = OraxenBlocks.getCustomBlockMechanic(blockBelow.getBlockData());
+                if (customBlockMechanic != null) yield customBlockMechanic.getInstrument().toLowerCase();
+                yield "block.note_block." + blockStateBelow.instrument().toString().toLowerCase();
+            }
+        };
+    }
```

Comment:
not necessary to be in nms classes, can just do it via a normal VersionUtil#atOrAbove / below checks

Reply:
How would I do that? I can't import CraftBlock or nms blockstate outside of the nms classes

Reply:
I see no reason you need nms blockstates for this at all
You can just do this via API

Reply:
I looked in the spigot javadocs and I couldn't find any method that would be able to return the instrument that a block would emit when played on top of, all the other methods would just return the value of the current noteblock's blockstate instrument value which is always gonna return HARP if u have noteblock blockstates updates disabled

Reply:
If there is no alternative, which im sure it is as i partially did this in my Blocky plugin iirc
just a small map as most blocks dont change the instrument but
You can streamline this to have all the non-nms stuff in one method and just an override for the blockstate
since ur repeating mostly the same stuff except the last line

Reply:
Yeah in the few recent commits I made it so it only overrides for the blockstate.
I dont think there is an alternative, I looked everywhere, making a map for each block's instrument isn't really an option I dont think

Reply:
https://github.com/MineInAbyss/Blocky/blob/master/src/main/kotlin/com/mineinabyss/blocky/helpers/NoteBlockHelpers.kt#L55-L92

this was my solution in another plugin, might be missing some modern blocks but it is taken from the wiki
i could swap to nms there though since well it relies on nms but

thats an API approach though it requires manually adding new blocks sure
nms is fine just abstract out all the non nms stuff 👍